## VAMOS A PROBAR ESTE DATASET DE GIT QUE HE ENCONTRADO: 


In [58]:
import pandas as pd
import logging
import numpy as np
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

route1 = '../data/SBS_Certified_Business_List.csv'  
route2 = '../data/bussinesses_licences.csv'  
route3 = '../data/Listing_of_Active_Businesses.csv'



try:  #me gusta
    df1 = pd.read_csv(route1)
    first_row = df1.iloc[0].to_dict()
    logging.info("First row df1:")
    for columna, valor in first_row.items():
        logging.info(f"{columna}: {valor}")
except Exception as e:
    logging.error("Error loading csv: %s", e)
    
print(df1.shape)
'''
try: # 2012
    df2 = pd.read_csv(route2)
    first_row = df2.iloc[0].to_dict()
    logging.info("First row df2:")
    for columna, valor in first_row.items():
        logging.info(f"{columna}: {valor}")
except Exception as e:
    logging.error("Error loading csv: %s", e)
    
print(df2.shape)

try: #very unbalanced, a lot of nans in naics
    df3 = pd.read_csv(route3)
    first_row = df3.iloc[0].to_dict()
    logging.info("First row df3:")
    for columna, valor in first_row.items():
        logging.info(f"{columna}: {valor}")
except Exception as e:
    logging.error("Error loading csv: %s", e)
    
print(df3.shape)


naics_nan_count = df3['NAICS'].isna().sum()
print(f"Número de filas con NAICS = NaN: {naics_nan_count}")
'''


2024-04-03 12:41:28,422 - INFO - First row df1:
2024-04-03 12:41:28,424 - INFO - Account_Number: 311147
2024-04-03 12:41:28,425 - INFO - Vendor_Formal_Name: #1Pho Inc
2024-04-03 12:41:28,426 - INFO - Vendor_DBA: Zenyai
2024-04-03 12:41:28,427 - INFO - First_Name: Albert
2024-04-03 12:41:28,428 - INFO - Last_Name: Jethanamest
2024-04-03 12:41:28,429 - INFO - telephone: 6463879761
2024-04-03 12:41:28,430 - INFO - Business_Description: Zenyai Viet Cajun & Pho Restaurant is dedicated to offering real Vietnamese flavor through distinct seafood boils and pho noodle dishes.
2024-04-03 12:41:28,431 - INFO - Certification: MBE
2024-04-03 12:41:28,432 - INFO - Certification_Renewal_Date: 10/31/2025
2024-04-03 12:41:28,433 - INFO - Ethnicity: ASIAN
2024-04-03 12:41:28,433 - INFO - Address_Line_1: 208 Grand Street
2024-04-03 12:41:28,435 - INFO - Address_Line_2: nan
2024-04-03 12:41:28,436 - INFO - City: Brooklyn
2024-04-03 12:41:28,440 - INFO - State: NY
2024-04-03 12:41:28,441 - INFO - Postcode:

(11430, 56)


'\ntry: # 2012\n    df2 = pd.read_csv(route2)\n    first_row = df2.iloc[0].to_dict()\n    logging.info("First row df2:")\n    for columna, valor in first_row.items():\n        logging.info(f"{columna}: {valor}")\nexcept Exception as e:\n    logging.error("Error loading csv: %s", e)\n    \nprint(df2.shape)\n\ntry: #very unbalanced, a lot of nans in naics\n    df3 = pd.read_csv(route3)\n    first_row = df3.iloc[0].to_dict()\n    logging.info("First row df3:")\n    for columna, valor in first_row.items():\n        logging.info(f"{columna}: {valor}")\nexcept Exception as e:\n    logging.error("Error loading csv: %s", e)\n    \nprint(df3.shape)\n\n\nnaics_nan_count = df3[\'NAICS\'].isna().sum()\nprint(f"Número de filas con NAICS = NaN: {naics_nan_count}")\n'

LET'S MAKE SOME ANALYSIS OF THE DATASET TO UNDERSTAND IT BETTER:

Firstly, let's clarify that we are carrying this project as a classification problem; where we are going to procced classifying each digit of the NAICS code one by one.

First digit of the NAICS code: Represents the general economic sector.
Second digit: Provides subsector information within the economic sector.
Third digit: Gives more details on the subsector's subdivision.
Fourth digit: Indicates the group industry.
Fifth digit: Offers even more specific information, often related to the national industry.
Sixth digit: Used for more detailed and specific classifications within a national industry, although not all countries use the sixth digit.


This approach results in a separate classification problem for each of the six digits of the NAICS code.

Before starting to solve any datset trade-offs, let's see how are the bussinesses on it for df1:

In [59]:

try:
    unique_naics_count = df1['ID6_digit_NAICS_code'].nunique()
    logging.info(f"Number of unique NAICS codes: {unique_naics_count}")

    top_repeated_naics = df1['ID6_digit_NAICS_code'].value_counts().head(1)
    logging.info("Top 5 most repeated NAICS codes and their counts:")
    for code, count in top_repeated_naics.items():
        logging.info(f"NAICS code {code} is repeated {count} times.")
    naics_counts = df1['ID6_digit_NAICS_code'].value_counts()
    ranges = [(1,20),(20, 50), (50, 100), (100, 200), (200, 300), (300, 400), (400, 500), (500, 600)]
    for r in ranges:
        count = naics_counts[(naics_counts >= r[0]) & (naics_counts <= r[1])].count()
        logging.info(f"Number of NAICS codes with occurrences between {r[0]} and {r[1]}: {count}")
except KeyError as e:
    logging.error(f"Column not found in DataFrame: {e}")
except Exception as e:
    logging.error(f"An unexpected error occurred during the analysis: {e}")
'''
try:
    unique_naics_count = df3['NAICS'].nunique()
    logging.info(f"Number of unique NAICS codes: {unique_naics_count}")

    top_repeated_naics = df3['NAICS'].value_counts().head(1)
    logging.info("Top 5 most repeated NAICS codes and their counts:")
    for code, count in top_repeated_naics.items():
        logging.info(f"NAICS code {code} is repeated {count} times.")
    naics_counts = df3['NAICS'].value_counts()
    ranges = [(1,20),(20, 50), (50, 100), (100, 200), (200, 300), (300, 400), (400, 500), (500, 600)]
    for r in ranges:
        count = naics_counts[(naics_counts >= r[0]) & (naics_counts <= r[1])].count()
        logging.info(f"Number of NAICS codes with occurrences between {r[0]} and {r[1]}: {count}")
except KeyError as e:
    logging.error(f"Column not found in DataFrame: {e}")
except Exception as e:
    logging.error(f"An unexpected error occurred during the analysis: {e}")
'''

2024-04-03 12:41:32,204 - INFO - Number of unique NAICS codes: 588
2024-04-03 12:41:32,208 - INFO - Top 5 most repeated NAICS codes and their counts:
2024-04-03 12:41:32,209 - INFO - NAICS code 541990 is repeated 534 times.
2024-04-03 12:41:32,212 - INFO - Number of NAICS codes with occurrences between 1 and 20: 480
2024-04-03 12:41:32,214 - INFO - Number of NAICS codes with occurrences between 20 and 50: 57
2024-04-03 12:41:32,216 - INFO - Number of NAICS codes with occurrences between 50 and 100: 29
2024-04-03 12:41:32,218 - INFO - Number of NAICS codes with occurrences between 100 and 200: 16
2024-04-03 12:41:32,220 - INFO - Number of NAICS codes with occurrences between 200 and 300: 3
2024-04-03 12:41:32,223 - INFO - Number of NAICS codes with occurrences between 300 and 400: 6
2024-04-03 12:41:32,226 - INFO - Number of NAICS codes with occurrences between 400 and 500: 0
2024-04-03 12:41:32,228 - INFO - Number of NAICS codes with occurrences between 500 and 600: 1


'\ntry:\n    unique_naics_count = df3[\'NAICS\'].nunique()\n    logging.info(f"Number of unique NAICS codes: {unique_naics_count}")\n\n    top_repeated_naics = df3[\'NAICS\'].value_counts().head(1)\n    logging.info("Top 5 most repeated NAICS codes and their counts:")\n    for code, count in top_repeated_naics.items():\n        logging.info(f"NAICS code {code} is repeated {count} times.")\n    naics_counts = df3[\'NAICS\'].value_counts()\n    ranges = [(1,20),(20, 50), (50, 100), (100, 200), (200, 300), (300, 400), (400, 500), (500, 600)]\n    for r in ranges:\n        count = naics_counts[(naics_counts >= r[0]) & (naics_counts <= r[1])].count()\n        logging.info(f"Number of NAICS codes with occurrences between {r[0]} and {r[1]}: {count}")\nexcept KeyError as e:\n    logging.error(f"Column not found in DataFrame: {e}")\nexcept Exception as e:\n    logging.error(f"An unexpected error occurred during the analysis: {e}")\n'

As of the latest revision, the American NAICS system defines approximately 1,057 distinct 6-digit codes. As we can see, in this dataset we have only defined 588 different NAICS codes. 
The range of NAICS codes in our dataset is influenced by the timing of data collection, reflecting economic conditions and industry relevance at that moment. Specific focus on certain industry subsectors or objectives may limit the diversity of NAICS codes captured. Geographic location could play a crucial role, as regional economic activities dictate the presence of particular industries. 

More precisely, we see that there are some NAICS codes that are repeated a lot of times. This is very important and significant since it probably means that the dataset is skewed towards certain industries or sectors. This skewness could indicate a concentration of economic activity within these sectors or a particular focus of the dataset collection efforts. Understanding this concentration can provide insights into industry dominance, regional economic strengths, or the specific research objectives guiding the data compilation.





In [50]:
nan_count = df3.isna().sum()
logging.info("missing values per feature:\n%s", nan_count)

2024-04-03 12:17:13,563 - INFO - missing values per feature:
LOCATION ACCOUNT #                0
BUSINESS NAME                     0
DBA NAME                     387790
STREET ADDRESS                    3
CITY                             53
ZIP CODE                          0
LOCATION DESCRIPTION             25
MAILING ADDRESS              285224
MAILING CITY                 285220
MAILING ZIP CODE             285327
NAICS                         64082
PRIMARY NAICS DESCRIPTION     64082
COUNCIL DISTRICT                  0
LOCATION START DATE            3574
LOCATION END DATE            585157
LOCATION                      29509
dtype: int64


We can get a few conclutions from this approach. 
As always, one of the reasons for NaNs can be errors in data collection, so I'm going to ommit this in the following explanations: 
    - vendor_DBA has a lot of Nans. Why? Maybe the formal name is the same as the Vendor_DBA. We can see in the following cell that even from those rows where Vendor_DBA has a value, 51 coincide with the formal name.

Let's firstly get rid of those features which are all NANs:

In [37]:
df= df.dropna(axis=1, how='all')
matching_rows = df[df['Vendor_DBA'].notna() & (df['Vendor_DBA'] == df['Vendor_Formal_Name'])].shape[0]
logging.info(f"Number of rows where 'Vendor_DBA' is not NaN and matches 'Vendor_Formal_Name': {matching_rows}")




2024-04-03 09:32:35,293 - INFO - Number of rows where 'Vendor_DBA' is not NaN and matches 'Vendor_Formal_Name': 51


Número de códigos NAICS distintos: 588
Máximo número de veces que un código NAICS se repite: 534
